### PR TITLE
Allow inclusion of JARs

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by mlutze** [comment]
@magnus-madsen does this resemble what you were thinking?

What's needed for the interface? Just a command line flag and some LSP stuff copied from adding fpkgs?

**Comment by magnus-madsen** [comment]
Yes, something like this. Then you would specify the jars on the command line, right?

**Comment by magnus-madsen** [comment]
Thus keeping the Flix compiler classpath separate from the JARs used in the project.

**Comment by magnus-madsen** [comment]
> 
> 
> Thus keeping the Flix compiler classpath separate from the JARs used in the project.

Or alternatively, the build system will suck in everything in lib and call addJar on the Flix object.

**Comment by magnus-madsen** [comment]
Ill add some comments tomorrow. Does it work as expected?

**Comment by magnus-madsen** [comment]
What happens if a JAR is corrupted / compiled with a too new version of Java?

**Comment by mlutze** [comment]
> Ill add some comments tomorrow. Does it work as expected?

Yes, with my sample size of 1 project :)



> What happens if a JAR is corrupted / compiled with a too new version of Java?

I would guess that the class loader would reject it. Worth testing.

**Comment by magnus-madsen** [comment]
> 
> 
> > Ill add some comments tomorrow. Does it work as expected?
> 
> Yes, with my sample size of 1 project :)
> 
> > What happens if a JAR is corrupted / compiled with a too new version of Java?
> 
> I would guess that the class loader would reject it. Worth testing.

My indirect question is whether the call in Resolver needs to catch extra exceptions. (and/or use of addJar)

File: main/src/ca/uwaterloo/flix/api/Flix.scala [hunk 1]
```diff
@@ -248,6 +249,11 @@ class Flix {
     */
   val genSym = new GenSym()
 
+  /**
+    * A class loader for loading JARs.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/Flix.scala:253]: => loading external jars

(I am going to suggest the word "external" to mean JARS that are not part of the Flix classpath.)

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala [hunk 2]
```diff
@@ -133,6 +139,7 @@ class LanguageServer(port: Int) extends WebSocketServer(new InetSocketAddress("l
     * Invoked when a client sends a message.
     */
   override def onMessage(ws: WebSocket, data: String): Unit = try {
+    Console.err.println(data) // MATT hacking logging
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:142]: remove

File: main/src/ca/uwaterloo/flix/api/lsp/Request.scala [hunk 3]
```diff
@@ -54,6 +54,16 @@ object Request {
     */
   case class RemPkg(requestId: String, uri: String) extends Request
 
+  /**
+    * A request to add (or update) the JAR at the given uri with the given binary data.
+    */
+  case class AddJar(requestId: String, uri: String) extends Request
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/Request.scala:60]: I assume if the jar is *updated* the class loader will not reload it. Probably a minor problem, but the comment is wrong.

File: main/src/ca/uwaterloo/flix/api/lsp/Request.scala [hunk 4]
```diff
@@ -183,6 +193,34 @@ object Request {
...
+  /**
+    * Tries to parse the given `json` value as a [[AddJar]] request.
+    */
+  def parseAddJar(json: json4s.JValue): Result[Request, String] = {
+    try {
+      for {
+        id <- parseId(json)
+        uri <- parseUri(json)
+      } yield {
+        val decoder = Base64.getDecoder
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/Request.scala:206]: decoder?

File: main/src/ca/uwaterloo/flix/language/phase/Namer.scala [hunk 5]
```diff
@@ -1493,8 +1493,8 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
     * Returns the class reflection object for the given `className`.
     */
   // TODO: Deprecated should be moved to resolver.
-  private def lookupClass(className: String, loc: SourceLocation): Validation[Class[_], NameError] = try {
-    Class.forName(className).toSuccess
+  private def lookupClass(className: String, loc: SourceLocation)(implicit flix: Flix): Validation[Class[_], NameError] = try {
+    Class.forName(className, true, flix.classLoader).toSuccess
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Namer.scala:1497]: Explain what is the Boolean (and why must be it be true?). I would assume that false would be better here?

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 6]
```diff
@@ -1969,16 +1969,16 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
   /**
     * Returns the class reflection object for the given `className`.
     */
-  private def lookupJvmClass(className: String, loc: SourceLocation): Validation[Class[_], ResolutionError] = try {
-    Class.forName(className).toSuccess
+  private def lookupJvmClass(className: String, loc: SourceLocation)(implicit flix: Flix): Validation[Class[_], ResolutionError] = try {
+    Class.forName(className, true, flix.classLoader).toSuccess
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:1973]: Same comment as in the other file. Explain why and maybe make false?

File: main/src/ca/uwaterloo/flix/util/AppendableClassLoader.scala [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.util
+
+import java.net.{URL, URLClassLoader}
+
+/**
+  * A class loader to which JARs can be added dynamically.
+  */
+class AppendableClassLoader extends URLClassLoader(Array.empty) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/util/AppendableClassLoader.scala:23]: Rename to `ExternalJarLoader`.

File: main/src/ca/uwaterloo/flix/util/AppendableClassLoader.scala [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+  /**
+    * Adds the URL to the class loader.
+    */
+  override def addURL(url: URL): Unit = {
+    // just reimplements the superclass, but makes it public
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/util/AppendableClassLoader.scala:29]: Good

**Review by magnus-madsen — COMMENTED** [review]
Overall looks good. Can you do an experiment.

Write a Java class with a static initializer (a static block) that prints something, compile it to a JAR, and put it in `lib` to see if it is loaded during compilation.

**Comment by mlutze** [comment]
> > > Ill add some comments tomorrow. Does it work as expected?
> > 
> > 
> > Yes, with my sample size of 1 project :)
> > > What happens if a JAR is corrupted / compiled with a too new version of Java?
> > 
> > 
> > I would guess that the class loader would reject it. Worth testing.
> 
> My indirect question is whether the call in Resolver needs to catch extra exceptions. (and/or use of addJar)

I tried with a corrupt jar (cut it in half). I just get a "Class not found" resolution error. Looking through the Javadoc for `UrlClassLoader`, it doesn't seem like `addUrl` or `getClass` has the possibility of throwing an exception relating to an invalid jar; it just skips over it.

**Comment by magnus-madsen** [comment]
What about the command line? Should it be possible to either specify `--jars a.jar b.jar` or just to give them directly `foo.flix bar.jar baz.flix` and have them handled correctly?

**Comment by magnus-madsen** [comment]
> > Overall looks good. Can you do an experiment.
> > Write a Java class with a static initializer (a static block) that prints something, compile it to a JAR, and put it in `lib` to see if it is loaded during compilation.
> 
> After changing those initialize flags to `false` as suggested, the static block is not executed.

OK and that is what we want, right?

**Comment by magnus-madsen** [comment]
> I tried with a corrupt jar (cut it in half). I just get a "Class not found" resolution error. Looking through the Javadoc for `UrlClassLoader`, it doesn't seem like `addUrl` or `getClass` has the possibility of throwing an exception relating to an invalid jar; it just skips over it.

OK, that is also good.

**Comment by mlutze** [comment]
> What about the command line? Should it be possible to either specify `--jars a.jar b.jar` or just to give them directly `foo.flix bar.jar baz.flix` and have them handled correctly?

My initial thought was that if a project is complex enough to need dependencies, then it should be a properly formatted project. But maybe that puts too much of a burden on the user?

> > > Overall looks good. Can you do an experiment.
> > > Write a Java class with a static initializer (a static block) that prints something, compile it to a JAR, and put it in `lib` to see if it is loaded during compilation.
> > 
> > 
> > After changing those initialize flags to `false` as suggested, the static block is not executed.
> 
> OK and that is what we want, right?

Right. No reason to execute code while compiling.

**Comment by magnus-madsen** [comment]
> > What about the command line? Should it be possible to either specify `--jars a.jar b.jar` or just to give them directly `foo.flix bar.jar baz.flix` and have them handled correctly?
> 
> My initial thought was that if a project is complex enough to need dependencies, then it should be a properly formatted project. But maybe that puts too much of a burden on the user?

Seems inconsistent. Why not support one of the two?

**Comment by mlutze** [comment]
Ok, I think putting the jars as options rather than arguments is best, since it implicitly communicates that just putting a jar is not a valid use of the command. Also, multi-argument options can be a pain sometimes. So I propose a repeatable `--jar` option.

**Comment by magnus-madsen** [comment]
> 
> 
> Ok, I think putting the jars as options rather than arguments is best, since it implicitly communicates that just putting a jar is not a valid use of the command. Also, multi-argument options can be a pain sometimes. So I propose a repeatable `--jar` option.

OK, let us do it.

But how will one provide multiple jars? Just separated by space?

**Comment by mlutze** [comment]
The command is repeatable, so you can do `--jar thing1.jar --jar thing2.jar`

**Comment by magnus-madsen** [comment]
> 
> 
> The command is repeatable, so you can do `--jar thing1.jar --jar thing2.jar`

OK, I definitely dont like that! Why would we have repeatable commands??? Can we just make it take multiple args?

**Comment by mlutze** [comment]
My understanding was that, by convention, an option has one or zero arguments. This prevents ambiguity between the arguments to the main command and the arguments to the option. (`--jars A B C D` which ones are jars and which are Flix files?)

We could pick a delimiter, but that makes things difficult if a file name contains that delimiter.

**Comment by magnus-madsen** [comment]
> 
> 
> My understanding was that, by convention, an option has one or zero arguments. This prevents ambiguity between the arguments to the main command and the arguments to the option. (`--jars A B C D` which ones are jars and which are Flix files?)
> 
> We could pick a delimiter, but that makes things difficult if a file name contains that delimiter.

How about we just allow an arbitrary mix of foo.flix and foo.jar? Like it seems pretty easy what to do. Just look at the extension?

**Comment by mlutze** [comment]
> > My understanding was that, by convention, an option has one or zero arguments. This prevents ambiguity between the arguments to the main command and the arguments to the option. (`--jars A B C D` which ones are jars and which are Flix files?)
> > We could pick a delimiter, but that makes things difficult if a file name contains that delimiter.
> 
> How about we just allow an arbitrary mix of foo.flix and foo.jar? Like it seems pretty easy what to do. Just look at the extension?

Yes, I guess that's the simplest solution

**Comment by mlutze** [comment]
Turns out it wasn't working as I thought it was. I get a `NoSuchFieldError`.

Flix file:
```scala
// The main entry point.
def main(_args: Array[String]): Int32 & Impure =
  import get Main:INT as getInt;
  Console.printLine("Hello World!");
  getInt() // exit code
```

Trace:
```
Exception in thread "main" java.lang.NoSuchFieldError: INT
        at Clo$main$114913.invoke(Unknown Source)
        at Ns.m_main(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:567)
        at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$link$1(JvmBackend.scala:277)
        at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$getCompiledMain$2(JvmBackend.scala:238)
        at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$getCompiledMain$2$adapted(JvmBackend.scala:236)
        at ca.uwaterloo.flix.Main$.main(Main.scala:207)
        at ca.uwaterloo.flix.Main.main(Main.scala)
```

Any pointers? I've tried messing with `FlixClassLoader` to no avail so far.

**Comment by magnus-madsen** [comment]
Yes, uh. I think the FlixClassLoader also has to know about these jars.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/FlixClassLoader.scala [hunk 9]
```diff
@@ -43,7 +44,8 @@ class FlixClassLoader(classes: Map[String, JvmClass]) extends ClassLoader {
         classes.get(name) match {
           case None =>
             // Case 1.1: The internal name does not exist. Try the super loader.
-            super.findClass(name)
+            flix.classLoader.loadClass(name) // MATT testing idk
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/FlixClassLoader.scala:47]: I was thinking something along these lines too...

**Comment by mlutze** [comment]
Turns out my testing file was messing things up since the name was `Main.java`, which conflicted with the `Main` class generated by the compiler.

The static initializer is indeed run at runtime; I think that's the right behavior.

Any thoughts on testing? The current import tests rely on classes that are actually in the compiler.

**Comment by magnus-madsen** [comment]
> Turns out my testing file was messing things up since the name was `Main.java`, which conflicted with the `Main` class generated by the compiler.
> 
> The static initializer is indeed run at runtime; I think that's the right behavior.
> 
> Any thoughts on testing? The current import tests rely on classes that are actually in the compiler.

Hmm. I think we can do without tests, once it works, its hard to see how it would break again. And if it does, we would quickly find out imo.

**Comment by magnus-madsen** [comment]
RFR?

**Comment by mlutze** [comment]
> RFR?

RFR

File: main/src/ca/uwaterloo/flix/language/phase/Namer.scala [hunk 10]
```diff
@@ -1493,8 +1493,10 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
     * Returns the class reflection object for the given `className`.
     */
   // TODO: Deprecated should be moved to resolver.
-  private def lookupClass(className: String, loc: SourceLocation): Validation[Class[_], NameError] = try {
-    Class.forName(className).toSuccess
+  private def lookupClass(className: String, loc: SourceLocation)(implicit flix: Flix): Validation[Class[_], NameError] = try {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Namer.scala:1496]: Remind me again why we have lookupClass in both Namer and Resolver?

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala [hunk 11]
```diff
@@ -292,6 +308,11 @@ class LanguageServer(port: Int) extends WebSocketServer(new InetSocketAddress("l
       }
     }
 
+    // Add JARs.
+    for (uri <- jars) {
+      flix.addJar(uri)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:313]: I think there could be other places where this is need too? Can you check where else `addInput` is called in `LanguageServer`? Or where `sources` is used?

File: main/src/ca/uwaterloo/flix/Main.scala [hunk 12]
```diff
@@ -179,7 +179,15 @@ object Main {
     val flix = new Flix()
     flix.setOptions(options)
     for (file <- cmdOpts.files) {
-      flix.addPath(file.toPath)
+      val ext = file.getName.split('.').last
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/Main.scala:182]: What happens if there is no dot? Crash?

File: main/src/ca/uwaterloo/flix/Main.scala [hunk 13]
```diff
@@ -179,7 +179,15 @@ object Main {
...
+      ext match {
+        case "flix" => flix.addPath(file.toPath)
+        case "fpkg" => flix.addPath(file.toPath)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/Main.scala:185]: I wonder if this should be refactored, but that could be in another PR.

**Comment by magnus-madsen** [comment]
Other than ensuring that we are using addJar in all the right places, this looks pretty reasonable I would say!

Do you have any comments/concerns?

**Comment by magnus-madsen** [comment]
What about in `Shell`? What happens there?

**Comment by magnus-madsen** [comment]
I guess for tests we don't care.

File: main/src/ca/uwaterloo/flix/Main.scala (hunk 12 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/Main.scala:182]: Split then returns a singleton array with an empty string.

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/Main.scala:182]: The full string *

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/Main.scala:182]: OK

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/Main.scala:182]: So .last is always defined.

File: main/src/ca/uwaterloo/flix/language/phase/Namer.scala (hunk 10 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Namer.scala:1496]: This is only used for catch cases. I think we need a bit of refactoring to avoid having to do this here.

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala (hunk 11 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:313]: I believe I got all the LSP places. ButI have to check `Shell` again

File: main/src/ca/uwaterloo/flix/Main.scala (hunk 13 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/Main.scala:185]: Yes, we need some file type checking utils, since there's some repeated code in a number of places.

File: main/src/ca/uwaterloo/flix/language/phase/Namer.scala (hunk 10 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Namer.scala:1496]: OK, can you create a ticket?

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala (hunk 11 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:313]: OK, let me know when done, and then I will merge.

File: main/src/ca/uwaterloo/flix/language/phase/Namer.scala (hunk 10 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Namer.scala:1496]: https://github.com/flix/flix/issues/2616

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala (hunk 11 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:313]: Interactive mode now supports jars.

**Comment by magnus-madsen** [comment]
Ready for merge?

**Comment by mlutze** [comment]
Ready!

**Comment by magnus-madsen** [comment]
@stephentetley This might be of interest to you. I think it should allow you to use a stock Flix compiler. 

(Maybe you will still need/want to use a nightly build).

**Comment by magnus-madsen** [comment]
Merged. Thanks! This is a nice feature to finally have.
